### ISSUE TITLE
Protobuf size explosion for TFLite

### ISSUE THREAD

**Issue by nestle1993** [tensorflow/tensorflow#17540]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04.3 LTS
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: tf 1.6.0
- **Python version**: Python 3.5.2
- **Bazel version (if compiling from source)**: bazel release 0.7.0
- **GCC/Compiler version (if compiling from source)**: gcc 4.9.4
- **CUDA/cuDNN version**: /
- **GPU model and memory**: /
- **Exact command to reproduce**: 
`bazel run --config=opt tensorflow/contrib/lite/toco:toco -- --input_file=/tmp/gru_rnn.pb --input_format=TENSORFLOW_GRAPHDEF  --input_types=FLOAT --output_format=TFLITE --output_file=/tmp/gru_rnn.tflite --inference_type=FLOAT --inference_input_type=FLOAT --output_arrays=model/softmax,model/rnn_states`

### Describe the problem
I successfully built a tflite model for mobile usage on November, 2017, and the size of .tflite was almost the same as .pb (no quantized), which was bout 950k.

However, when I try to convert the same .pb to tflite model with tensorflow 1.6 master branch (HEAD commit f7acdf2ed5e0b9c50c1c5f4b80163255aa9e8073), I find that the size of the .tflite file is 17M, **which is 17x larger than before!!**

So I make a binary search of the commits in `tensorflow/contrib/lite` history, then I find out at which commit that thing changes: the commit 90ce80131a8b5213d9f3eb9649d63921db7874a4 can produce 950k .tflite model, but right after it 528d0c5e4d148655b797368fd55fe6304730fece will make the tflite model become 17M.

Here are some additional information:
  - Both of these two tflite models can be successfully run on mobile.
  - I did not check the commits (if there have) between these two, since I only trace the `/tensorflow/contrib/lite` history.
  - I guess the problem happened due to some of the operations I used in my frozen graph, in which I have 2 layer gru and some linear transforms. But I cannot upload the .pb file here...

**Comment by andrehentz** [comment]
Thanks! We are investigating.

**Comment by andrehentz** [comment]
@nestle1993  A couple of questions for you:
   - do you have a BatchMatMul in your pb?
   - if so, what is its shape?

**Comment by andrehentz** [comment]
@nestle1993 It maybe that the generated file has lots of zeros in it (something we are trying to fix). It would be nice to confirm this is the case. 

If you looked at your flatbuffer in a hex editor and sees tons of 0s that would indicate that being the problem. Or zip it up - if it gets 17x smaller that's probably a good indicator too.

thanks!

**Comment by nestle1993** [comment]
@andrehentz 
I think I don't have BatchMatMul op in my frozen graph, at least I don't use it explicitly.
I open the .pb file with the hex editor, there are 960+ thousand lines in total, and 40 thousand lines are 0s at the beginning of it.
Then I open the .tflite file, there seems no many 0s.. 

For making the debug easier, I have uploaded these three files (original .pb, 950k .tflie and 17M .tflite) here:
https://drive.google.com/file/d/13IJHGc9HJ7iS0_fRE9OYvuFoxBMjOYVW/view?usp=sharing
https://drive.google.com/file/d/1mWWPRkA5z3KaQ2m-adE9k2HEGHs6Um5b/view?usp=sharing
https://drive.google.com/file/d/1ETqhCC9Bwor65F6zkS23FCQrI8hdA_DZ/view?usp=sharing

thanks!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by xiaoyunwu** [comment]
@andrehentz any update?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by xiaoyunwu** [comment]
hello? @andrehentz

**Comment by kaierlong** [comment]
I also met this problem when I try to transfer lightrnn model which is protobuf format to tflite. And my original protobuf model's size is 9M, but in tflite format the model is bigger than 100M.

**Comment by andrehentz** [comment]
Sorry for the long delay. The fix for this issue will be made available in the next few days (likely early next week).

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 23 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 37 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @andrehentz: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by anitha-v** [comment]
Closing it , please reopen if the issue persists

**Comment by jibanreloaded** [comment]
I am also facing the same issue. created a stack overflow issue on the same. tensorflow lite model after conversion is around 245MB where as the original tensorflow mobile is around ~1MB.

https://stackoverflow.com/questions/52039880/tensorflow-lite-conversion-for-lstm-model

**Comment by aselle** [comment]
Thanks for the report @jibanreloaded, we will look into this problem as well.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by andrehentz** [comment]
Here are my results based on a recent build of tflite_convert, using the gru_rnn.pb file from above.
```
$ bazel run //tensorflow/lite/python:tflite_convert -- --graph_def_file /tmp/gru_rnn.pb --output_file /tmp/gru_rnn.tflite --output_arrays=model/softmax,model/rnn_states --input_arrays=model/frame_0,model/frame_1,model/frame_2,model/frame_3,model/frame_4,model/frame_5,model/frame_6,model/frame_7,model/frame_8,model/frame_9,model/frame_10,model/frame_11,model/frame_12,model/frame_13,model/frame_14,model/frame_15,model/frame_16,model/frame_17,model/frame_18,model/frame_19,model/frame_20,model/frame_21,model/frame_22,model/rnn_initial_state_0,model/rnn_initial_state_1

$ ls -sh /tmp/gru_rnn.tflite
848k /tmp/as.tflite

```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jibanreloaded** [comment]
> Here are my results based on a recent build of tflite_convert, using the gru_rnn.pb file from above.
> 
> ```
> $ bazel run //tensorflow/lite/python:tflite_convert -- --graph_def_file /tmp/gru_rnn.pb --output_file /tmp/gru_rnn.tflite --output_arrays=model/softmax,model/rnn_states --input_arrays=model/frame_0,model/frame_1,model/frame_2,model/frame_3,model/frame_4,model/frame_5,model/frame_6,model/frame_7,model/frame_8,model/frame_9,model/frame_10,model/frame_11,model/frame_12,model/frame_13,model/frame_14,model/frame_15,model/frame_16,model/frame_17,model/frame_18,model/frame_19,model/frame_20,model/frame_21,model/frame_22,model/rnn_initial_state_0,model/rnn_initial_state_1
> 
> $ ls -sh /tmp/gru_rnn.tflite
> 848k /tmp/as.tflite
> ```

@andrehentz : which version of tensorflow are you using and did you try the same with toco?

**Comment by andrehentz** [comment]
I'm building from HEAD. Here's the equivalent toco command line:

`blaze run //third_party/tensorflow/lite/toco:toco -- --input_file=/tmp/gru_rnn.pb --output_file=/tmp/gru_rnn.tflite --output_arrays=model/softmax,model/rnn_states --input_arrays=model/frame_0,model/frame_1,model/frame_2,model/frame_3,model/frame_4,model/frame_5,model/frame_6,model/frame_7,model/frame_8,model/frame_9,model/frame_10,model/frame_11,model/frame_12,model/frame_13,model/frame_14,model/frame_15,model/frame_16,model/frame_17,model/frame_18,model/frame_19,model/frame_20,model/frame_21,model/frame_22,model/rnn_initial_state_0,model/rnn_initial_state_1`

**Comment by suharshs** [comment]
Closing due to inactivity.
